### PR TITLE
Add checks before hoisting out in loop pipelining 

### PR DESCRIPTION
**Pull request by fotiskoun**
Currently, during a loop pipelining transformation, operations may be hoisted out without any checks on the loop bounds, which leads to incorrect transformations and unexpected behaviour. The following [issue ](https://github.com/llvm/llvm-project/issues/90870) describes the problem more extensively, including an example.
The proposed fix adds some check in the loop bounds before and applies the maximum hoisting.

### CODE REVIEW

**Comment by fotiskoun** [comment]
Ping

File: mlir/lib/Dialect/SCF/TransformOps/SCFTransformOps.cpp [hunk 1]
```diff
@@ -217,6 +217,8 @@ loopScheduling(scf::ForOp forOp,
     return 1;
   };
 
+  auto ubConstant = getConstantIntValue(forOp.getUpperBound());
```
Comment (ftynse) [inline mlir/lib/Dialect/SCF/TransformOps/SCFTransformOps.cpp:220]: Nit: please expand `auto` unless the type is obvious from line-level context or impossible to spell out.

File: mlir/lib/Dialect/SCF/TransformOps/SCFTransformOps.cpp [hunk 2]
```diff
@@ -227,7 +229,14 @@ loopScheduling(scf::ForOp forOp,
       Operation *def = operand.getDefiningOp();
       if (!def)
         continue;
-      earlyCycle = std::max(earlyCycle, opCycles[def] + getLatency(def));
+      if (ubConstant && lbConstant) {
+        unsigned ubInt = ubConstant.value();
+        unsigned lbInt = lbConstant.value();
+        auto minLatency = std::min(ubInt - lbInt - 1, getLatency(def));
+        earlyCycle = std::max(earlyCycle, opCycles[def] + minLatency);
+      } else {
```
Comment (ftynse) [inline mlir/lib/Dialect/SCF/TransformOps/SCFTransformOps.cpp:237]: How should this work when loop bounds are dynamic? Could we have a test?

**Review by ftynse — COMMENTED** [review]
Thanks for the patch

File: mlir/lib/Dialect/SCF/TransformOps/SCFTransformOps.cpp (hunk 1 above)
Comment (fotiskoun) [reply mlir/lib/Dialect/SCF/TransformOps/SCFTransformOps.cpp]: Thanks, I changed `auto` to `std::optional<int64_t>`

**Comment by fotiskoun** [comment]
@ftynse I have addressed the comments. Thank you!

**Review by ftynse — APPROVED** [review]

**Pull request merged**